### PR TITLE
Compound exchange rate

### PR DESCRIPTION
**Pull request by mds1**
Modifies Compound trigger from checking an invariant to checking that exchange rate only increases. That gas costs of not checking trigger condition multiple times per block (i.e. short circuit the check) were investigated and determined not to be worth it. Details:

- If we short circuit when block number is the same, subsequent calls save about 9k gas per call. But this isn’t ideal because I could exploit Compound in [block 1, tx1], see that exploit, then borrow max from Cozy in [block 1, tx2], then have my debts cancelled in block 2 when trigger is toggled
- If we short circuit when block number AND exchange rate are both the same, the first call in a block is 13,434 gas and subsequent calls are 10,258 gas
- If we don’t short circuit, the first call in a block is 13,143 gas (291 less gas) and subsequent calls are 10,343 gas (85 more gas)
- Conclusion here is that there's no clear benefits to short-circuiting, so lets keep it simple and check every time to be safe

(Note: the reason for this PR is that the previous invariant approach was flawed because it'd only be valid for a very brief period. For example if someone stole cash from a contract, it could be triggered, but as soon as someone interacts with Compound the invariant would be "fixed" because it relies on an updated (but now broken) exchange rate and now cannot be triggered)

Question—for naming convention, do we like the current format of `<Protocol><Condition>`, i.e. `CompoundExchangeRate`, or should it be something like `CompoundExchangeRateTrigger`, or something else?

### CODE REVIEW

File: contracts/CompoundExchangeRate.sol [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * @dev Checks the Compound Invariant that reserves + supply = cash + borrows
+   */
+  function isMarketTriggered() internal override returns (bool) {
+    // Read this blocks exchange rate
+    uint256 _currentExchangeRate = market.exchangeRateStored();
+
+    // Check if current exchange rate is below current exchange rate, accounting for tolerance
+    bool _status = _currentExchangeRate < lastExchangeRate - tolerance;
```
Comment (apbendi) [inline contracts/CompoundExchangeRate.sol]: Recommend adding parens for readability:

```solidity
bool _status = _currentExchangeRate < (lastExchangeRate - tolerance);
```

**Review by apbendi — APPROVED** [review]
Looks good. Agree with the conclusions on short circuiting. One nit for readability.

File: contracts/CompoundExchangeRate.sol (hunk 1 above)
Comment (mds1) [reply contracts/CompoundExchangeRate.sol]: Done in f89a2511d4485cb80e954f3b925a447a1124e56a

**Comment by mds1** [comment]
Holding off on merging for now as we investigate a suitable error tolerance

**Pull request merged**